### ISSUE TITLE
compiling fails for boost 1.58

### ISSUE THREAD

**Issue by aschuch247** [freeorion/freeorion#3172]
Bug Report
==========

The compilation of FreeOrion 4ce1758 from source code fails.

```
/freeorion/GG/src/dialogs/ColorDlg.cpp: In member function ‘virtual void GG::ColorDlg::CompleteConstruction()’:
/freeorion/GG/src/dialogs/ColorDlg.cpp:610:18: error: ‘boost::placeholders’ has not been declared
     using boost::placeholders::_1;
                  ^
/freeorion/GG/src/dialogs/ColorDlg.cpp:611:18: error: ‘boost::placeholders’ has not been declared
     using boost::placeholders::_2;
                  ^
/freeorion/GG/src/dialogs/ColorDlg.cpp:612:18: error: ‘boost::placeholders’ has not been declared
     using boost::placeholders::_3;
                  ^
/freeorion/GG/src/dialogs/FileDlg.cpp: In member function ‘virtual void GG::FileDlg::CompleteConstruction()’:
/freeorion/GG/src/dialogs/FileDlg.cpp:202:27: error: ‘placeholders’ is not a namespace-name
     namespace ph = boost::placeholders;
                           ^
/freeorion/GG/src/dialogs/FileDlg.cpp:202:39: error: expected namespace-name before ‘;’ token
     namespace ph = boost::placeholders;
                                       ^
/freeorion/GG/src/dialogs/FileDlg.cpp:206:92: error: ‘ph’ has not been declared
     m_files_list->SelRowsChangedSignal.connect(boost::bind(&FileDlg::FileSetChanged, this, ph::_1));
                                                                                            ^
/freeorion/GG/src/dialogs/FileDlg.cpp:207:97: error: ‘ph’ has not been declared
     m_files_list->DoubleClickedRowSignal.connect(boost::bind(&FileDlg::FileDoubleClicked, this, ph::_1, ph::_2, ph::_3));
                                                                                                 ^
/freeorion/GG/src/dialogs/FileDlg.cpp:207:105: error: ‘ph’ has not been declared
     m_files_list->DoubleClickedRowSignal.connect(boost::bind(&FileDlg::FileDoubleClicked, this, ph::_1, ph::_2, ph::_3));
                                                                                                         ^
/freeorion/GG/src/dialogs/FileDlg.cpp:207:113: error: ‘ph’ has not been declared
     m_files_list->DoubleClickedRowSignal.connect(boost::bind(&FileDlg::FileDoubleClicked, this, ph::_1, ph::_2, ph::_3));
                                                                                                                 ^
/freeorion/GG/src/dialogs/FileDlg.cpp:208:86: error: ‘ph’ has not been declared
     m_files_edit->EditedSignal.connect(boost::bind(&FileDlg::FilesEditChanged, this, ph::_1));
                                                                                      ^
/freeorion/GG/src/dialogs/FileDlg.cpp:209:88: error: ‘ph’ has not been declared
     m_filter_list->SelChangedSignal.connect(boost::bind(&FileDlg::FilterChanged, this, ph::_1));
                                                                                        ^
/freeorion/universe/Fleet.cpp: In function ‘std::__cxx11::list<int> {anonymous}::TruncateRouteToEndAtSystem(const std::__cxx11::list<int>&, int, int)’:
/freeorion/universe/Fleet.cpp:70:85: error: ‘boost::placeholders’ has not been declared
                                    boost::bind(&SystemHasNoVisibleStarlanes, boost::placeholders::_1, empire_id));
                                                                                     ^
/freeorion/universe/Conditions.cpp:51:18: error: ‘boost::placeholders’ has not been declared
     using boost::placeholders::_1;
                  ^
```

Environment
-----------

* **FreeOrion Version**: 4ce1758
* **Operating System**: Ubuntu 16.04 LTS
* **Fetched as**
  * Compiled from source
* Boost 1.58
* GCC (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609

Description
-----------

The source code does not compile, see above.

Expected Result
---------------

The source code shall compile.

Steps to reproduce
------------------

* mkdir build
* cd build
* ccmake ..
* make

**Comment by o01eg** [comment]
Does Ubuntu16.04 LTS support boost-python3? We moved testing to 18.04 because of lack python3 support.

**Comment by aschuch247** [comment]
@o01eg: Do you know what package I shall look for?

**Comment by aschuch247** [comment]
Package: libboost-python1.58-dev
Version: 1.58.0+dfsg-5ubuntu3.1
Depends: libboost1.58-dev (= 1.58.0+dfsg-5ubuntu3.1), libboost-python1.58.0 (= 1.58.0+dfsg-5ubuntu3.1), python-dev, python:any (<< 2.8), python:any (>= 2.7.5-5~), python (>= 2.5) | python-celementtree | python-elementtree

I guess not then?

**Comment by o01eg** [comment]
No, cmake shouldn't be able to generate make file.

Could you show what you have in `cat CMakeCache.txt | grep -i python` ?

**Comment by aschuch247** [comment]
```
$ cat CMakeCache.txt | grep -i python
//Boost python35 library (debug)
Boost_PYTHON35_LIBRARY_DEBUG:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_python-py35.so
//Boost python35 library (release)
Boost_PYTHON35_LIBRARY_RELEASE:FILEPATH=/usr/lib/x86_64-linux-gnu/libboost_python-py35.so
//Python tool for Style Guide Enforcement
PYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3.5
PYTHON_INCLUDE_DIR:PATH=/usr/include/python3.5m
PYTHON_LIBRARY:FILEPATH=/usr/lib/x86_64-linux-gnu/libpython3.5m.so
PYTHON_LIBRARY_DEBUG:FILEPATH=PYTHON_LIBRARY_DEBUG-NOTFOUND
PYTHON_LIBRARY_RELEASE:FILEPATH=PYTHON_LIBRARY_RELEASE-NOTFOUND
//ADVANCED property for variable: Boost_PYTHON35_LIBRARY_DEBUG
Boost_PYTHON35_LIBRARY_DEBUG-ADVANCED:INTERNAL=1
//ADVANCED property for variable: Boost_PYTHON35_LIBRARY_RELEASE
Boost_PYTHON35_LIBRARY_RELEASE-ADVANCED:INTERNAL=1
//Details about finding PythonInterp
FIND_PACKAGE_MESSAGE_DETAILS_PythonInterp:INTERNAL=[/usr/bin/python3.5][v3.5.2(3.5)]
//Details about finding PythonLibs
FIND_PACKAGE_MESSAGE_DETAILS_PythonLibs:INTERNAL=[/usr/lib/x86_64-linux-gnu/libpython3.5m.so][/usr/include/python3.5m][v3.5.2(3.5)]
//ADVANCED property for variable: PYTHON_EXECUTABLE
PYTHON_EXECUTABLE-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_INCLUDE_DIR
PYTHON_INCLUDE_DIR-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY
PYTHON_LIBRARY-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY_DEBUG
PYTHON_LIBRARY_DEBUG-ADVANCED:INTERNAL=1
//ADVANCED property for variable: PYTHON_LIBRARY_RELEASE
PYTHON_LIBRARY_RELEASE-ADVANCED:INTERNAL=1
_Boost_COMPONENTS_SEARCHED:INTERNAL=atomic;chrono;date_time;filesystem;iostreams;locale;log;log_setup;python35;regex;serialization;system;thread
```

**Comment by geoffthemedio** [comment]
Looks like the oldest Boost that has the `_1` style placeholders in `namespace boost::placeholders` is 1.60:

https://www.boost.org/doc/libs/1_60_0/boost/bind/placeholders.hpp
https://www.boost.org/doc/libs/1_59_0/boost/bind/placeholders.hpp

If it's only needed for systems that don't support Python3 anyway, then presumably it's reasonable to leave as "won't fix" and bump the minimum required Boost version?

**Comment by o01eg** [comment]
> If it's only needed for systems that don't support Python3 anyway, then presumably it's reasonable to leave as "won't fix" and bump the minimum required Boost version?

@geoffthemedio No, looks like it supports boost with Python3.5.

**Comment by geoffthemedio** [comment]
Does this actually fit v0.4.10 or just master?

**Comment by o01eg** [comment]
0.4.10 have the same code https://github.com/freeorion/freeorion/blob/cb6ef2c1d985a59bccb9c37c49a5dc62ccf64da6/GG/src/dialogs/ColorDlg.cpp#L617

**Comment by o01eg** [comment]
Should we drop 1.58?

**Comment by o01eg** [comment]
@aschuch247 Could you write in `ColorDlg.cpp`

```cpp
#if BOOST_VERSION >= 106000
    using boost::placeholders::_1;
    using boost::placeholders::_2;
    using boost::placeholders::_3;
#endif
```

Will it help?

**Comment by geoffthemedio** [comment]
There are lots of places that use those placeholders... would be a bit more substantial to update than just that one usage.

**Comment by o01eg** [comment]
It was introduced here: https://github.com/freeorion/freeorion/pull/2959

**Comment by aschuch247** [comment]
> ```c++
> #if BOOST_VERSION >= 106000
>     using boost::placeholders::_1;
>     using boost::placeholders::_2;
>     using boost::placeholders::_3;
> #endif
> ```

@o01eg: Yes, I now have a `ColorDlg.cpp.o`.

**Comment by o01eg** [comment]
Wonder if make it for all files in #2959 allows to restore Ubuntu 16.04 LTS compatibility.

**Comment by geoffthemedio** [comment]
> Wonder if make it for all files in #2959 allows to restore Ubuntu 16.04 LTS compatibility.

might be some more in https://github.com/freeorion/freeorion/pull/3123

**Comment by o01eg** [comment]
Somehow some placeholders left and works (until 1.74): https://github.com/freeorion/freeorion/blob/4ce1758820e12d2cb0592c11cf8cec5ead362f13/UI/GovernmentWnd.cpp#L596

**Comment by geoffthemedio** [comment]
those resolve to `boost::placeholders::_1` and equivalent. not sure why.

**Comment by o01eg** [comment]
@aschuch247 It looks like it's decided to bump boost version so could you update Ubuntu to 18.04 LTS with supported boost?

**Comment by geoffthemedio** [comment]
We can still adapt for 1.58. Just until that happens, only 1.60+ work...

**Comment by Vezzra** [comment]
@geoffthemedio, 

> We can still adapt for 1.58. Just until that happens, only 1.60+ work...

But for 0.4.10.1, compatibility with boost 1.60+ should be sufficient, right? In which case we can assign this to 0.5?

Otherwise RC3 would become necessary...

**Comment by o01eg** [comment]
I've tried to fix code to be compatible with boost 1.58 but found some other issue on Xenial 16.04

```
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[23]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
             {MeterType::METER_SPEED,          ClientUI::MeterIcon(MeterType::METER_SPEED),     "FW_FLEET_SPEED_SUMMARY"}};
                                                                                                            ^
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[24]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[25]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[23]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[24]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[26]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[26]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[27]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[27]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[24]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[22]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
/freeorion/UI/FleetWnd.cpp:1722:108: error: converting to 'std::tuple<MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {MeterType, std::shared_ptr<GG::Texture>, const char (&)[23]}; <template-parameter-2-2> = void; _Elements = {MeterType, std::shared_ptr<GG::Texture>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}]'
```

```
g++ --version
g++ (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
```

**Comment by o01eg** [comment]
And next supposed to be related to initializer list as well: 

```
In file included from /usr/include/c++/5/bits/hashtable.h:35:0,
                 from /usr/include/c++/5/unordered_map:47,
                 from /freeorion/GG/GG/Font.h:24,
                 from /freeorion/GG/GG/TextControl.h:22,
                 from /freeorion/GG/GG/Button.h:23,
                 from /freeorion/UI/CUIWnd.h:5,
                 from /freeorion/UI/EncyclopediaDetailPanel.h:6,
                 from /freeorion/UI/EncyclopediaDetailPanel.cpp:1:
/usr/include/c++/5/bits/hashtable_policy.h: In instantiation of 'struct std::__detail::__is_noexcept_hash<PlanetType, std::hash<PlanetType> >':
/usr/include/c++/5/type_traits:137:12:   required from 'struct std::__and_<std::__is_fast_hash<std::hash<PlanetType> >, std::__detail::__is_noexcept_hash<PlanetType, std::hash<PlanetType> > >'
/usr/include/c++/5/type_traits:148:38:   required from 'struct std::__not_<std::__and_<std::__is_fast_hash<std::hash<PlanetType> >, std::__detail::__is_noexcept_hash<PlanetType, std::hash<PlanetType> > > >'
/usr/include/c++/5/bits/unordered_map.h:100:65:   required from 'class std::unordered_map<PlanetType, std::vector<std::__cxx11::basic_string<char> > >'
/freeorion/UI/EncyclopediaDetailPanel.cpp:2730:73:   required from here
/usr/include/c++/5/bits/hashtable_policy.h:85:34: error: no match for call to '(const std::hash<PlanetType>) (const PlanetType&)'
  noexcept(declval<const _Hash&>()(declval<const _Key&>()))>
```

**Comment by geoffthemedio** [comment]
> And next supposed to be related to initializer list as well:

That's not a boost issue, is it?

rather: https://stackoverflow.com/questions/18837857/cant-use-enum-class-as-unordered-map-key

But when do you see this error? all the tests succeed...

**Comment by o01eg** [comment]
I declared `std::hash` for `PlanetType` and it compiled.

**Comment by geoffthemedio** [comment]
On that linked page it says that it should have been fixed in C++14. What standard are you using?

**Comment by o01eg** [comment]
It uses C++14 but looks like gcc 5.4 didn't support it.

**Comment by geoffthemedio** [comment]
Is it just that one `unordered_map` that has the issue? If so, then maybe add the `hash` declaration within an `#if` testing for GCC and the relevant version just where needed...

"officially" we require C++14, so a non-conformant compiler is not supported, so I'd implement a minimal fix at most. In Ubuntu 16, can you not install a newer GCC?

I'd actually like to move to C++17 soon if possible.

**Comment by o01eg** [comment]
> In Ubuntu 16, can you not install a newer GCC?

No, they only updated 5.3 to 5.4.

**Comment by o01eg** [comment]
> If so, then maybe add the hash declaration within an #if testing for GCC and the relevant version just where needed...

It doesn't break something so I think it fine to left it unguarded.

> I'd actually like to move to C++17 soon if possible.

I think we can fix compatibility for 0.4.10 and left 0.5+ as is.

**Comment by geoffthemedio** [comment]
> I think we can fix compatibility for 0.4.10 and left 0.5+ as is.

Yes, that's fine, if we specifically want to target Ubuntu 16 for v0.4.10. LTS ends next year, I think I read, so no need to go out of the way to support GCC < 6 in master.
